multipart-post.lua

```diff
@@ -36,44 +40,111 @@ local gen_boundary = function()
   return table.concat(t)
 end
 
-local encode = function(t, boundary)
-    boundary = boundary or gen_boundary()
+local encode = function(r, k, v, boundary)
+    local _t = type(v)
+
+    tprintf(r, "--%s\r\n", boundary)
+    if _t == "string" then
+        section_header(r, k, {})
+    elseif _t == "table" then
+        assert(v.data, "invalid input")
+        local extra = {
+            filename = v.filename or v.name,
+            content_type = v.content_type or v.mimetype
+                or "application/octet-stream",
+            content_transfer_encoding = v.content_transfer_encoding or "binary",
+        }
+        section_header(r, k, extra)
+    else
+        error(string.format("unexpected type %s", _t))
+    end
+end
+
+local encode_source = function(k, v, boundary)
+    r = {}
+    encode(r, k, v, boundary)
+    return ltn12.source.string(table.concat(r))
+end
+
+local data_len = function(d)
+    local _t = type(d)
+
+    if _t == "string" then
+        return string.len(d)
+    elseif _t == "table" then
+        if type(d.data) == "string" then
+            return string.len(d.data)
+        end
+        if d.len then return d.len end
+        error("must provide data length for non-string datatypes")
+    end
+end
+
+local content_length = function(t, boundary)
     local r = {}
-    local _t
-    for k,v in pairs(t) do
-        tprintf(r, "--%s\r\n", boundary)
-        _t = type(v)
+    local data_length = 0
+    for k, v in pairs(t) do
+        encode(r, k, v, boundary)
```

Comment:
This PR means you encode everything twice now. There should be a way to store those lengths in a context so it isn't necessary to recompute them.

Reply:
Will fix after merging.